From 223fd84588909beae1f19440e4668ac5580f9f71 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Fri, 19 Sep 2014 12:29:17 -0400
Subject: [acpi] hotplug/pci: Execute _EJ0 under the ACPI scan lock

Message-id: <1411129776-21656-43-git-send-email-prarit@redhat.com>
Patchwork-id: 96389
O-Subject: [RHEL7.1 PATCH BZ 1141846 42/61] ACPI / hotplug / PCI: Execute _EJ0 under the ACPI scan lock
Bugzilla: 1141846
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Myron Stowe <mstowe@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1141846

commit 21369c77477a7f937174833c8094154f0f995710
Author: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
Date:   Mon Feb 10 13:36:26 2014 +0100

    ACPI / hotplug / PCI: Execute _EJ0 under the ACPI scan lock

    Since acpi_device_hotplug() assumes that ACPI handles of device
    objects passed to it will not become invalid while acpi_scan_lock
    is being held, make acpiphp_disable_slot() acquire acpi_scan_lock,
    because it generally causes _EJ0 to be executed for one of the
    devices in the slot and that may cause its ACPI handle to become
    invalid.

    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Myron Stowe <mstowe@redhat.com>
Cc: Andy Gospodarek <agospoda@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/pci/hotplug/acpiphp_glue.c b/drivers/pci/hotplug/acpiphp_glue.c
index 5d45a40..c35447e 100644
--- a/drivers/pci/hotplug/acpiphp_glue.c
+++ b/drivers/pci/hotplug/acpiphp_glue.c
@@ -1007,9 +1007,15 @@ int acpiphp_disable_slot(struct acpiphp_slot *slot)
 {
  int ret;
 
+ /*
+  * Acquire acpi_scan_lock to ensure that the execution of _EJ0 in
+  * acpiphp_disable_and_eject_slot() will be synchronized properly.
+  */
+ acpi_scan_lock_acquire();
  pci_lock_rescan_remove();
  ret = acpiphp_disable_and_eject_slot(slot);
  pci_unlock_rescan_remove();
+ acpi_scan_lock_release();
  return ret;
 }
 
-- 
1.7.1